File: tests/bm25.test.ts
import { describe, expect, test } from "bun:test";
import { BM25Index } from "../src/index/bm25.ts";

describe("BM25Index", () => {
  test("ranks matching document higher than unrelated document", () => {
    const index = new BM25Index();
    index.index([
      ["auth", "middleware", "token"],
      ["database", "migration", "schema"],
    ]);
    const scores = index.getScores(["auth", "token"]);
    expect(scores[0]).toBeGreaterThan(scores[1] ?? 0);
  });

  test("ranks document with more query terms higher", () => {
    const index = new BM25Index();
    index.index([["alpha"], ["alpha", "beta", "gamma"], ["delta", "epsilon"]]);
    const scores = index.getScores(["alpha", "beta"]);
    expect(scores[1]).toBeGreaterThan(scores[0] ?? 0);
    expect(scores[1]).toBeGreaterThan(scores[2] ?? 0);
  });

  test("returns zero scores for unknown query terms", () => {
    const index = new BM25Index();
    index.index([["known", "term"]]);
    const scores = index.getScores(["missing", "terms"]);
    expect(scores.every((s) => s === 0)).toBe(true);
  });

  test("respects weight mask", () => {
    const index = new BM25Index();
    index.index([
      ["auth", "token"],
      ["auth", "token", "extra"],
    ]);
    const masked = index.getScores(["auth"], [false, true]);
    expect(masked[0]).toBe(0);
    expect(masked[1]).toBeGreaterThan(0);
  });

  test("getScoresAsync matches getScores", async () => {
    const docs: string[][] = [];
    for (let d = 0; d < 300; d++) {
      docs.push(
        d % 2 === 0 ? ["alpha", "beta", "gamma", `doc${d}`] : ["delta", "epsilon", `other${d}`],
      );
    }
    const index = new BM25Index();
    index.index(docs);
    const query = ["alpha", "doc42", "epsilon"];
    expect(await index.getScoresAsync(query)).toEqual(index.getScores(query));
  });
});

File: internal/index/bm25_test.go
package index

import (
	"reflect"
	"strconv"
	"testing"
)

func TestBM25RanksMatchingDocumentHigherThanUnrelatedDocument(t *testing.T) {
	idx := NewBM25Index()
	idx.Index([][]string{
		{"auth", "middleware", "token"},
		{"database", "migration", "schema"},
	})
	scores := idx.GetScores([]string{"auth", "token"})
	if scores[0] <= scores[1] {
		t.Fatalf("scores = %#v, want matching document higher", scores)
	}
}

func TestBM25RanksDocumentWithMoreQueryTermsHigher(t *testing.T) {
	idx := NewBM25Index()
	idx.Index([][]string{{"alpha"}, {"alpha", "beta", "gamma"}, {"delta", "epsilon"}})
	scores := idx.GetScores([]string{"alpha", "beta"})
	if scores[1] <= scores[0] || scores[1] <= scores[2] {
		t.Fatalf("scores = %#v, want doc with more query terms higher", scores)
	}
}

func TestBM25ReturnsZeroScoresForUnknownQueryTerms(t *testing.T) {
	idx := NewBM25Index()
	idx.Index([][]string{{"known", "term"}})
	scores := idx.GetScores([]string{"missing", "terms"})
	for _, score := range scores {
		if score != 0 {
			t.Fatalf("scores = %#v, want all zero", scores)
		}
	}
}

func TestBM25RespectsWeightMask(t *testing.T) {
	idx := NewBM25Index()
	idx.Index([][]string{{"auth", "token"}, {"auth", "token", "extra"}})
	masked := idx.GetScores([]string{"auth"}, []bool{false, true})
	if masked[0] != 0 || masked[1] <= 0 {
		t.Fatalf("scores = %#v, want first masked and second positive", masked)
	}
}

func TestBM25GetScoresAsyncMatchesGetScores(t *testing.T) {
	docs := [][]string{}
	for d := 0; d < 300; d++ {
		if d%2 == 0 {
			docs = append(docs, []string{"alpha", "beta", "gamma", "doc" + strconv.Itoa(d)})
		} else {
			docs = append(docs, []string{"delta", "epsilon", "other" + strconv.Itoa(d)})
		}
	}
	idx := NewBM25Index()
	idx.Index(docs)
	query := []string{"alpha", "doc42", "epsilon"}
	if !reflect.DeepEqual(idx.GetScoresAsync(query), idx.GetScores(query)) {
		t.Fatal("GetScoresAsync does not match GetScores")
	}
}

func TestBM25JSONRoundTrip(t *testing.T) {
	idx := NewBM25Index()
	idx.Index([][]string{{"alpha", "alpha"}, {"beta"}})
	loaded := BM25FromJSON(idx.ToJSON())
	if !reflect.DeepEqual(loaded.GetScores([]string{"alpha"}), idx.GetScores([]string{"alpha"})) {
		t.Fatal("BM25 JSON round-trip changed scores")
	}
}
